crates/uv-auth/src/keyring.rs

```diff
@@ -33,77 +38,160 @@ pub enum Error {
     ParseFailed(#[from] std::string::FromUtf8Error),
 }
 
-/// Get credentials from keyring for given url
-///
-/// See `pip`'s KeyringCLIProvider
-/// <https://github.com/pypa/pip/blob/ae5fff36b0aad6e5e0037884927eaa29163c0611/src/pip/_internal/network/auth.py#L102>
-pub fn get_keyring_subprocess_auth(url: &Url) -> Result<Option<Credential>, Error> {
-    let host = url.host_str();
-    if host.is_none() {
-        return Err(Error::NotKeyringTarget(
-            "Should only use keyring for urls with host".to_string(),
-        ));
+impl KeyringProvider {
+    /// Fetch credentials for the given [`Url`] from the keyring.
+    ///
+    /// Returns `None` if no password was found, even if a username is present on the URL.
+    pub(crate) fn fetch(&self, url: &Url) -> Result<Option<Credentials>, Error> {
+        let host = url.host_str();
+        if host.is_none() {
+            return Err(Error::NotKeyringTarget(
+                "Should only use keyring for urls with host".to_string(),
+            ));
+        }
+
+        if url.password().is_some() {
+            return Err(Error::NotKeyringTarget(
+                "Url already contains password - keyring not required".to_string(),
+            ));
+        }
+
+        debug!("Checking keyring for credentials for `{}`", url.to_string(),);
```

Comment:
The display for `Url` is not in the same format as the string, the string is better for a user-facing message.

Reply:
Sorry, what’s the difference?

Reply:
`to_string` is just a wrapper around `Display`:

```rust
impl<T: fmt::Display + ?Sized> ToString for T {
    // A common guideline is to not inline generic functions. However,
    // removing `#[inline]` from this method causes non-negligible regressions.
    // See <https://github.com/rust-lang/rust/pull/74852>, the last attempt
    // to try to remove it.
    #[inline]
    default fn to_string(&self) -> String {
        let mut buf = String::new();
        let mut formatter = core::fmt::Formatter::new(&mut buf);
        // Bypass format_args!() to avoid write_str with zero-length strs
        fmt::Display::fmt(self, &mut formatter)
            .expect("a Display implementation returned an error unexpectedly");
        buf
    }
}
```

Could you be confusing `Display` and `Debug`?

Reply:
Um... weird. I swear I was getting different output. Let me check back on that.

Reply:
(It's possible that types can override this behavior and maybe it _is_ different for URL, I just assumed it wasn't.)

Reply:
I think I just got confused by some of the reqwest internal logs which look like `("http", 127.0.0.1:53514)`